File: WorkLog/Services/TimerService.swift
import Foundation
import Observation

@MainActor
protocol TimerServiceProtocol: AnyObject {
    var activeSession: Session? { get }
    var activeProject: Project? { get }
    var elapsed: TimeInterval { get }
    var isRunning: Bool { get }

    func start(project: Project) throws
    func pause() throws
    func resume() throws
    func stop() throws
    func addManualSession(project: Project, startTime: Date, endTime: Date, note: String) throws
}

@MainActor
@Observable
final class TimerService: TimerServiceProtocol {
    private(set) var activeSession: Session?
    private(set) var activeProject: Project?
    private(set) var elapsed: TimeInterval = 0

    var isRunning: Bool { activeSession != nil }

    private let sessionRepository: SessionRepositoryProtocol
    private let projectRepository: ProjectRepositoryProtocol
    private let validationService: ValidationServiceProtocol
    private let idleDetectionService: IdleDetectionServiceProtocol
    private var tickTask: Task<Void, Never>?

    init(
        sessionRepository: SessionRepositoryProtocol,
        projectRepository: ProjectRepositoryProtocol,
        validationService: ValidationServiceProtocol = ValidationService(),
        idleDetectionService: IdleDetectionServiceProtocol
    ) {
        self.sessionRepository = sessionRepository
        self.projectRepository = projectRepository
        self.validationService = validationService
        self.idleDetectionService = idleDetectionService

        restoreActiveSessionIfNeeded()

        self.idleDetectionService.onShouldAutoPause = { [weak self] in
            try? self?.pause()
        }
    }

    private func restoreActiveSessionIfNeeded() {
        guard let session = try? sessionRepository.fetchActiveSession() else { return }
        activeSession = session
        activeProject = session.project
        elapsed = Date.now.timeIntervalSince(session.startTime)
        startTicking()
    }

    func start(project: Project) throws {
        try closeCurrentRunningSession(status: .paused)

        let session = Session(
            project: project,
            date: .now,
            startTime: .now,
            category: project.category,
            status: .running
        )
        try sessionRepository.insert(session)

        activeSession = session
        activeProject = project
        elapsed = 0
        startTicking()
    }

    func pause() throws {
        try closeCurrentRunningSession(status: .paused)
        stopTicking()
    }

    func resume() throws {
        guard let lastProject = try sessionRepository.fetchRecent(limit: 1).first?.project else { return }
        try start(project: lastProject)
    }

    func stop() throws {
        try closeCurrentRunningSession(status: .completed)
        stopTicking()
    }

    func addManualSession(project: Project, startTime: Date, endTime: Date, note: String) throws {
        try validationService.validateSessionFields(project: project, startTime: startTime, endTime: endTime)
        guard try !sessionRepository.hasOverlap(start: startTime, end: endTime, excluding: nil) else {
            throw ValidationError.overlappingSession
        }
        let session = Session(
            project: project,
            date: startTime,
            startTime: startTime,
            endTime: endTime,
            durationSeconds: endTime.timeIntervalSince(startTime),
            note: note,
            category: project.category,
            status: .completed
        )
        try sessionRepository.insert(session)
    }

    private func closeCurrentRunningSession(status: SessionStatus) throws {
        guard let session = activeSession else { return }
        let now = Date.now
        session.endTime = now
        session.durationSeconds = now.timeIntervalSince(session.startTime)
        session.status = status
        try sessionRepository.update(session)
        activeSession = nil
        activeProject = nil
        elapsed = 0
    }

    private func startTicking() {
        tickTask?.cancel()
        tickTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self, let session = self.activeSession else { return }
                self.elapsed = Date.now.timeIntervalSince(session.startTime)
                try? await Task.sleep(for: .seconds(1))
            }
        }
    }

    private func stopTicking() {
        tickTask?.cancel()
        tickTask = nil
    }
}

File: desktop-electron/electron/services/timerService.ts
import { BrowserWindow, powerMonitor } from "electron";
import type { Project, Session, SessionInput, TimerState } from "../../src/shared/types";
import { secondsBetween } from "../../src/shared/format";
import { validateSessionFields } from "../../src/shared/validation";
import { ProjectRepository } from "./projectRepository";
import { SessionRepository } from "./sessionRepository";
import { SettingsRepository } from "./settingsRepository";

export class TimerService {
  private tick: NodeJS.Timeout | null = null;

  constructor(
    private readonly projects: ProjectRepository,
    private readonly sessions: SessionRepository,
    private readonly settings: SettingsRepository,
    private readonly broadcast: () => void
  ) {
    powerMonitor.on("suspend", () => {
      this.pause().catch(() => undefined);
    });
    powerMonitor.on("lock-screen", () => {
      this.pause().catch(() => undefined);
    });
    setInterval(() => {
      const active = this.sessions.fetchActive();
      const idleLimit = this.settings.get().idleTimeoutMinutes * 60;
      if (active && powerMonitor.getSystemIdleTime() >= idleLimit) {
        this.pause().catch(() => undefined);
      }
    }, 30_000);
  }

  state(): TimerState {
    const activeSession = this.sessions.fetchActive();
    const activeProject = activeSession?.projectId ? this.projects.fetchById(activeSession.projectId) : null;
    const elapsedSeconds = activeSession ? secondsBetween(activeSession.startTime, new Date().toISOString()) : 0;
    this.ensureTick(Boolean(activeSession));
    return {
      activeSession,
      activeProject,
      elapsedSeconds,
      isRunning: Boolean(activeSession)
    };
  }

  async start(projectId: string): Promise<TimerState> {
    const project = this.projects.fetchById(projectId);
    if (!project) throw new Error("Projeto não encontrado.");
    this.closeCurrentRunningSession("paused");
    this.sessions.insert({
      projectId: project.id,
      date: new Date().toISOString(),
      startTime: new Date().toISOString(),
      endTime: null,
      durationSeconds: 0,
      note: "",
      category: project.category,
      status: "running"
    });
    this.broadcast();
    return this.state();
  }

  async pause(): Promise<TimerState> {
    this.closeCurrentRunningSession("paused");
    this.broadcast();
    return this.state();
  }

  async stop(): Promise<TimerState> {
    this.closeCurrentRunningSession("completed");
    this.broadcast();
    return this.state();
  }

  async resume(): Promise<TimerState> {
    const recent = this.sessions.fetchRecent(1)[0];
    if (!recent?.projectId) return this.state();
    return this.start(recent.projectId);
  }

  addManualSession(input: SessionInput): Session {
    const project = this.projects.fetchById(input.projectId);
    validateSessionFields(project, input);
    if (this.sessions.hasOverlap(input.startTime, input.endTime, null)) {
      throw new Error("Já existe uma sessão registrada nesse período.");
    }
    const session = this.sessions.insert({
      projectId: input.projectId,
      date: input.startTime,
      startTime: input.startTime,
      endTime: input.endTime,
      durationSeconds: secondsBetween(input.startTime, input.endTime),
      note: input.note.trim(),
      category: (project as Project).category,
      status: "completed"
    });
    this.broadcast();
    return session;
  }

  updateManualSession(id: string, input: SessionInput): Session {
    const project = this.projects.fetchById(input.projectId);
    validateSessionFields(project, input);
    if (this.sessions.hasOverlap(input.startTime, input.endTime, id)) {
      throw new Error("Já existe uma sessão registrada nesse período.");
    }
    const session = this.sessions.update(id, {
      projectId: input.projectId,
      date: input.startTime,
      startTime: input.startTime,
      endTime: input.endTime,
      durationSeconds: secondsBetween(input.startTime, input.endTime),
      note: input.note.trim(),
      category: (project as Project).category,
      status: "completed"
    });
    this.broadcast();
    return session;
  }

  registerWindow(window: BrowserWindow): void {
    window.on("closed", () => {
      if (this.tick && BrowserWindow.getAllWindows().length === 0) {
        clearInterval(this.tick);
        this.tick = null;
      }
    });
  }

  private closeCurrentRunningSession(status: "paused" | "completed"): void {
    const active = this.sessions.fetchActive();
    if (!active) return;
    const now = new Date().toISOString();
    this.sessions.update(active.id, {
      endTime: now,
      durationSeconds: secondsBetween(active.startTime, now),
      status
    });
  }

  private ensureTick(shouldRun: boolean): void {
    if (!shouldRun && this.tick) {
      clearInterval(this.tick);
      this.tick = null;
      return;
    }
    if (shouldRun && !this.tick) {
      this.tick = setInterval(this.broadcast, 1000);
    }
  }
}
